Task: Stateful Set force delete pods

foxish opened the pull request:
Task: Force Deleting a Pod

* Mentions when it is appropriate to do this.
* Cautions the user about the various risks of doing this.
* Gives the user the command to run.

cc @kubernetes/sig-apps @kow3ns @janetkuo @enisoc

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1731)
<!-- Reviewable:end -->

enisoc commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,68 @@
+
+#### Force Deletion
+
+Force deletions do **not** wait for confirmation from the kubelet that the pod has been terminated. Irrespective of whether a force deletion is successful in killing a pod, it will immediately free up the name from the apiserver. This would let the Stateful Set controller create a replacement pod with that same identity. However, the risks of doing so must be evident. If there is a pod running on ...
```
Elsewhere it seems like we're trying to avoid assuming users know what "fenced" means. Maybe elaborate a bit: "... is a user's assertion that the pod in question will never again make contact with other pods in the Set."

foxish replied:
I was planning to link this reference to "fenced" to the node controller documentation where we can explain it. But yes, at a glance, what you said seems like a good replacement.

kargakis replied:
Probably something for the debugging doc for StatefulSets but it would be nice if we could showcase such a scenario (unreachable node, pods can still communicate between them) and walk people through it (exec on a pod of the StatefulSet that is on a healthy node, do something that proves it can communicate with the partitioned pod).

foxish commented:
@erictune PTAL.

erictune commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,68 @@
+   * The node object is deleted (either by the user, or by the node controller).
+   * The kubelet on the unresponsive node starts responding, kills the pod and removes the entry from the apiserver. 
+   * Force deletion of the pod by the user.
+   
```
Maybe add:

It is preferred to use the first or second approaches.  If a node is confirmed to be dead (e.g. the network or power plug is removed, etc) then the node object should be deleted.)  If the node is suffering from a network partition, then try to resolve this or wait for it to resolve, and then the kubelet will complete the deletion of the pod.

erictune commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,68 @@
+   * The kubelet on the unresponsive node starts responding, kills the pod and removes the entry from the apiserver. 
+   * Force deletion of the pod by the user.
+   
+As explained above, there are cases in which graceful deletion will be unable to remove a pod. In these cases, if you want the Stateful Set controller to create a replacement pod with the same identity, you may want to perform force deletion.
```
I think the phrase `graceful deletion` is just meant to describe getting termination notice and waiting for terminationPeriodGracePeriod.  It doesn't encompass the node controller node deletion behavior, etc.  So, maybe phrase this as "Normally, pods are only fully deleted by the system (by kubelets or pod GC after node deletion) .   In the case when the system is not working as intended you may want to perform force deletion."

smarterclayton commented:
I would probably recommend against "fully deleted".  As I've tried to write
docs and CLI code around this, the terms that have come up that seemed
consistent:

1. "request deletion" - user sends a DELETE call
2. "waiting for deletion" - the period of time between the request and the
deletion
3. "force deletion" - user explicitly finishes the deletion

So I might say "Normally, the system completes the deletion once the pod is
no longer running on a node, or the node is deleted by an administrator.
You may override this by force deleting the pod."

On Sun, Nov 27, 2016 at 3:34 PM, Eric Tune <notifications@github.com> wrote:

> *@erictune* commented on this pull request.
>
> required review comment.
> ------------------------------
>
> In docs/tasks/stateful-sets/deleting-pods.md
> <https://github.com/kubernetes/kubernetes.github.io/pull/1731#pullrequestreview-10236522>
> :
>
> > +{% capture prerequisites %}
> +
> +* If you are not using Stateful Sets, you can skip this task now.
> +* This is a fairly advanced task and has the potential to violate some properties provided by Stateful Sets.
> +* If you are attempting this, you must be well aware of the various considerations enumerated below, and have specific domain knowledge of the application(s) running using Stateful Set.
> +
> +{% endcapture %}
> +
> +{% capture steps %}
> +
> +
> +### Stateful Set considerations
> +
> +In normal operation of a Stateful Set, there is **never** a need to force delete a Stateful Set pod. The Stateful Set controller is responsible for creating, scaling and deleting members of the Stateful Set. It tries to ensure that the specified number of pods from ordinal 0 through N-1 are alive and ready. Stateful Set also gives us the property that there is at any time at most one pod with a given identity running in a cluster. We will hereby refer to this as *at most one* semantics provided by a Stateful Set. (*TODO: Link to concept page.*)
> +
> +Manual force deletion is a maintenance task that must be undertaken with caution. It has the potential to violate *at-most-one* semantics provided by Stateful Sets. Stateful Sets may be used to run distributed and clustered applications which have a need for a stable network identity and stable storage. These applications often have configuration which relies on a an ensemble of a fixed number of members with fixed identities. Having multiple members with the same identity can be disastrous and may lead to data loss (e.g. split brain scenario in quorum-based systems).
>
> s/on a an/on an/
> ------------------------------
>
> In docs/tasks/stateful-sets/deleting-pods.md
> <https://github.com/kubernetes/kubernetes.github.io/pull/1731#pullrequestreview-10236522>
> :
>
> > +
> +Manual force deletion is a maintenance task that must be undertaken with caution. It has the potential to violate *at-most-one* semantics provided by Stateful Sets. Stateful Sets may be used to run distributed and clustered applications which have a need for a stable network identity and stable storage. These applications often have configuration which relies on a an ensemble of a fixed number of members with fixed identities. Having multiple members with the same identity can be disastrous and may lead to data loss (e.g. split brain scenario in quorum-based systems).
> +
> +### Deletion of Pods
> +
> +#### Graceful Deletion
> +
> +Graceful deletion may be performed using the following:
> +
> +```shell
> +$ kubectl delete pods <pod>
> +```
> +
> +For the above to lead to graceful termination, the pod **must** not specify a `pod.Spec.TerminationGracePeriodSeconds` of 0. The practice of setting a `pod.Spec.TerminationGracePeriodSeconds` of 0 seconds is unsafe and strongly discouraged for Stateful Set pods. Graceful deletion is safe and will ensure that the pod shuts down gracefully before the kubelet deletes the name from the apiserver (*TODO: link to http://kubernetes.io/docs/user-guide/pods/#termination-of-pods).
> +
> + In order for Stateful Sets to provide *at most one* semantics, changes were made to the Node Controller in 1.5, (*TODO: Link to Node Controller behavior page.*) so that it never force deletes pods on unreachable nodes. When a node becomes unreachable for an extended period of time, we deem such a node as having been partitioned. The pods running on an unreachable node enter the 'Terminating' or 'Unknown' state eventually. Pods may also enter these states when the user attempts graceful deletion (*link to kubectl documentation*) of a pod on a partitioned node. The only ways in which a pod in such a state can be removed from the apiserver are as follows:
>
> Avoid "we". Avoid writing a history. Say "Kubernetes (versions 1.5 or
> newer) will not delete pods just because a node is unreachable."
> ------------------------------
>
> In docs/tasks/stateful-sets/deleting-pods.md
> <https://github.com/kubernetes/kubernetes.github.io/pull/1731#pullrequestreview-10236522>
> :
>
> > +
> +#### Graceful Deletion
> +
> +Graceful deletion may be performed using the following:
> +
> +```shell
> +$ kubectl delete pods <pod>
> +```
> +
> +For the above to lead to graceful termination, the pod **must** not specify a `pod.Spec.TerminationGracePeriodSeconds` of 0. The practice of setting a `pod.Spec.TerminationGracePeriodSeconds` of 0 seconds is unsafe and strongly discouraged for Stateful Set pods. Graceful deletion is safe and will ensure that the pod shuts down gracefully before the kubelet deletes the name from the apiserver (*TODO: link to http://kubernetes.io/docs/user-guide/pods/#termination-of-pods).
> +
> + In order for Stateful Sets to provide *at most one* semantics, changes were made to the Node Controller in 1.5, (*TODO: Link to Node Controller behavior page.*) so that it never force deletes pods on unreachable nodes. When a node becomes unreachable for an extended period of time, we deem such a node as having been partitioned. The pods running on an unreachable node enter the 'Terminating' or 'Unknown' state eventually. Pods may also enter these states when the user attempts graceful deletion (*link to kubectl documentation*) of a pod on a partitioned node. The only ways in which a pod in such a state can be removed from the apiserver are as follows:
> +   * The node object is deleted (either by the user, or by the node controller).
> +   * The kubelet on the unresponsive node starts responding, kills the pod and removes the entry from the apiserver.
> +   * Force deletion of the pod by the user.
> +
>
> Maybe add:
>
> It is preferred to use the first or second approaches. If a node is
> confirmed to be dead (e.g. the network or power plug is removed, etc) then
> the node object should be deleted.) If the node is suffering from a network
> partition, then try to resolve this or wait for it to resolve, and then the
> kubelet will complete the deletion of the pod.
> ------------------------------
>
> In docs/tasks/stateful-sets/deleting-pods.md
> <https://github.com/kubernetes/kubernetes.github.io/pull/1731#pullrequestreview-10236522>
> :
>
> > +#### Graceful Deletion
> +
> +Graceful deletion may be performed using the following:
> +
> +```shell
> +$ kubectl delete pods <pod>
> +```
> +
> +For the above to lead to graceful termination, the pod **must** not specify a `pod.Spec.TerminationGracePeriodSeconds` of 0. The practice of setting a `pod.Spec.TerminationGracePeriodSeconds` of 0 seconds is unsafe and strongly discouraged for Stateful Set pods. Graceful deletion is safe and will ensure that the pod shuts down gracefully before the kubelet deletes the name from the apiserver (*TODO: link to http://kubernetes.io/docs/user-guide/pods/#termination-of-pods).
> +
> + In order for Stateful Sets to provide *at most one* semantics, changes were made to the Node Controller in 1.5, (*TODO: Link to Node Controller behavior page.*) so that it never force deletes pods on unreachable nodes. When a node becomes unreachable for an extended period of time, we deem such a node as having been partitioned. The pods running on an unreachable node enter the 'Terminating' or 'Unknown' state eventually. Pods may also enter these states when the user attempts graceful deletion (*link to kubectl documentation*) of a pod on a partitioned node. The only ways in which a pod in such a state can be removed from the apiserver are as follows:
> +   * The node object is deleted (either by the user, or by the node controller).
> +   * The kubelet on the unresponsive node starts responding, kills the pod and removes the entry from the apiserver.
> +   * Force deletion of the pod by the user.
> +
> +As explained above, there are cases in which graceful deletion will be unable to remove a pod. In these cases, if you want the Stateful Set controller to create a replacement pod with the same identity, you may want to perform force deletion.
>
> I think the phrase graceful deletion is just meant to describe getting
> termination notice and waiting for terminationPeriodGracePeriod. It doesn't
> encompass the node controller node deletion behavior, etc. So, maybe phrase
> this as "Normally, pods are only fully deleted by the system (by kubelets
> or pod GC after node deletion) . In the case when the system is not working
> as intended you may want to perform force deletion."
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes.github.io/pull/1731#pullrequestreview-10236522>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p_nikamfwAJlftyu3fI8kh9wXF8aks5rCelCgaJpZM4K27w->
> .
>

calebamiles commented:
How does this PR look @smarterclayton, @erictune I'd love to close out https://github.com/kubernetes/kubernetes/issues/36034 once this is merged. Thanks!

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+---
+
+{% capture overview %}
+This page shows how to delete pods which are part of a stateful set, and explains the various considerations when doing so.
```
"stateful set" -> "StatefulSet"

devin-donnelly replied:
"and explains the considerations to keep in mind when doing so."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+{% capture prerequisites %}
+
+* If you are not using Stateful Sets, you can skip this task now. 
```
This is not a prerequisite. Please remove.

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+{% capture prerequisites %}
+
+* If you are not using Stateful Sets, you can skip this task now. 
+* This is a fairly advanced task and has the potential to violate some properties provided by Stateful Sets.
```
"to violate some of the properties inherent to StatefulSet."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+* If you are not using Stateful Sets, you can skip this task now. 
+* This is a fairly advanced task and has the potential to violate some properties provided by Stateful Sets.
+* If you are attempting this, you must be well aware of the various considerations enumerated below, and have specific domain knowledge of the application(s) running using Stateful Set.
```
"Before proceeding, make yourself familiar with the considerations enumerated below,"

devin-donnelly replied:
Stateful Set -> StatefulSet; please change throughout.

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+### Stateful Set considerations
+
+In normal operation of a Stateful Set, there is **never** a need to force delete a Stateful Set pod. The Stateful Set controller is responsible for creating, scaling and deleting members of the Stateful Set. It tries to ensure that the specified number of Pods from ordinal 0 through N-1 are alive and ready. Stateful Set also gives us the property that there is at any time at most one pod with a g ...
```
Avoid first-person collective constructs ("we", "us"). Instead of "StatefulSet gives us the property", try "StatefulSet ensures that, at any time, there is at most one pod with a given identity running in a cluster."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+In normal operation of a Stateful Set, there is **never** a need to force delete a Stateful Set pod. The Stateful Set controller is responsible for creating, scaling and deleting members of the Stateful Set. It tries to ensure that the specified number of Pods from ordinal 0 through N-1 are alive and ready. Stateful Set also gives us the property that there is at any time at most one pod with a g ...
+
+Manual force deletion is a maintenance task that must be undertaken with caution. It has the potential to violate *at most one* semantics provided by Stateful Sets. Stateful Sets may be used to run distributed and clustered applications which have a need for a stable network identity and stable storage. These applications often have configuration which relies on an ensemble of a fixed number of m ...
```
A little wordy.

"Manual force deletion should be undertaken with caution, as it has the potential to violate the *at most one* semantics inherent to StatefulSet."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+Manual force deletion is a maintenance task that must be undertaken with caution. It has the potential to violate *at most one* semantics provided by Stateful Sets. Stateful Sets may be used to run distributed and clustered applications which have a need for a stable network identity and stable storage. These applications often have configuration which relies on an ensemble of a fixed number of m ...
+
+### Deletion of Pods
```
Try not to nest headings directly under one another; usually means the doc has a structural problem.

In this case, "Graceful Deletion" is not necessary, and you should change "Deletion of Pods" to "Deleting Pods" (the former is passive voice).

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+For the above to lead to graceful termination, the pod **must not** specify a `pod.Spec.TerminationGracePeriodSeconds` of 0. The practice of setting a `pod.Spec.TerminationGracePeriodSeconds` of 0 seconds is unsafe and strongly discouraged for Stateful Set Pods. Graceful deletion is safe and will ensure that the [Pod shuts down gracefully](/docs/user-guide/pods/#termination-of-pods) before the ku ...
+
+Kubernetes (versions 1.5 or newer) will not delete Pods just because a Node is unreachable. The Pods running on an unreachable Node enter the 'Terminating' or 'Unknown' state eventually. Pods may also enter these states when the user attempts graceful deletion (*link to kubectl documentation*) of a pod on an unreachable Node. The only ways in which a pod in such a state can be removed from the ap ...
```
Probably need to explain what "eventually" is; isn't the default five minutes?

devin-donnelly replied:
Do not refer to the reader as "user" in indirect third person. Address the reader directly as "you."

"Pods may also enter these states when you attempt to gracefully terminate a pod on an unreachable node."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+   * The kubelet on the unresponsive Node starts responding, kills the Pod and removes the entry from the apiserver. 
+   * Force deletion of the pod by the user.
+   
+It is preferred to use the first or second approaches. If a Node is confirmed to be dead (e.g. permanently disconnected from the network, powered down, etc) then the node object should be deleted. If the node is suffering from a network partition, then try to resolve this or wait for it to resolve. When the partition heals, the kubelet will complete the deletion of the pod and free up its name in ...
```
"The recommended best practice is to use the first or second approach."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+#### Force Deletion
+
+Force deletions do **not** wait for confirmation from the kubelet that the pod has been terminated. Irrespective of whether a force deletion is successful in killing a pod, it will immediately free up the name from the apiserver. This would let the Stateful Set controller create a replacement pod with that same identity. However, the risks of doing so must be evident. If there is a pod running on ...
```
I'm against "must be evident." We need to spell these things out.

"This would allow the StatefulSet controller to create a replacement pod with the same identity; this can lead to the duplication of a still-running pod, and if said pod can still communicate with the other members of the StatefulSet, will violate the *at most one* semantics that StatefulSet is designed to guarantee."

devin-donnelly replied:
New paragraph before the last sentence.

devin-donnelly commented on docs/tasks/stateful-sets/deleting-pods.md:
```diff
@@ -0,0 +1,70 @@
+
+Force deletions do **not** wait for confirmation from the kubelet that the pod has been terminated. Irrespective of whether a force deletion is successful in killing a pod, it will immediately free up the name from the apiserver. This would let the Stateful Set controller create a replacement pod with that same identity. However, the risks of doing so must be evident. If there is a pod running on ...
+
+If you want to delete a Pod forcibly, do the following.
```
Replace period after "following" with colon.

foxish commented:
@devin-donnelly Thanks for the detailed review. Addressed comments. PTAL. 

Preview url: https://deploy-preview-1731--kubernetes-io-vnext-staging.netlify.com/docs/tasks/manage-stateful-set/delete-pods/

devin-donnelly commented on docs/tasks/manage-stateful-set/delete-pods.md:
```diff
@@ -0,0 +1,75 @@
+
+### Deleting Pods
+
+Graceful deletion may be performed using the following:
```
"You can perform a graceful pod deletion with the following command:"

devin-donnelly commented on docs/tasks/manage-stateful-set/delete-pods.md:
```diff
@@ -0,0 +1,75 @@
+
+Force deletions **do not** wait for confirmation from the kubelet that the Pod has been terminated. Irrespective of whether a force deletion is successful in killing a Pod, it will immediately free up the name from the apiserver. This would let the StatefulSet controller create a replacement Pod with that same identity; this can lead to the duplication of a still-running Pod, and if said Pod can  ...
+
+Force deletion of a StatefulSet Pod is a user's assertion that the Pod in question will never again make contact with other Pods in the StatefulSet and its name can be safely freed up for a replacement to be created. 
```
We can make this have more impact by directly addressing the user:

"When you force delete a StatefulSet pod, you are asserting that the Pod in question will never again..."

devin-donnelly commented:
Just a couple more comments. Are the tech review comments also resolved? Should someone tech LGTM this?

foxish commented:
@devin-donnelly fixed. Thanks!

erictune commented:
LGTM.  Good balance of scaring the user into not force deleting, and telling them how and when they might want to do this.
